### PR TITLE
feat: Introduce NodeBuilder

### LINKED ISSUE #5693: Node Builder for e2e testing
### Describe the feature

Currently we have no easy way to create the node for end-to-end testing of RPC, and the only way to end-to-end test RPC is through hive. It should be easy to create e2e RPC + engine tests, to increase coverage of RPC endpoints that are currently not well covered. Tracing and engine endpoints are good candidates for e2e tests that need almost all parts of the node.

## Requirements
End-to-end tests require these main things:
* Simple process for launching a reth node
* Tools for generating and modifying test data outside of the node
  * For example, reorg tests will need main and side chains to be fed into the node at test time
* Finally, the test vectors themselves

This issue focuses on the first point, easily launching a reth node with real components. Testing should also make sure that any database or other file created is ephemeral, and only exists for the life of the test. This means we need to support `TempDatabase` as well as `DatabaseEnv` in node init methods.

## API Design
The [anvil `NodeConfig`](https://github.com/foundry-rs/foundry/blob/f5c91995f80b5bf3b4c29c934d414cc198c9e7a8/crates/anvil/src/config.rs#L91) and [`spawn`](https://github.com/foundry-rs/foundry/blob/f5c91995f80b5bf3b4c29c934d414cc198c9e7a8/crates/anvil/src/lib.rs#L96-L96) methods provide good inspiration for how the reth node builder should be used in tests. Anvil tests use a [`NodeHandle`](https://github.com/foundry-rs/foundry/blob/f5c91995f80b5bf3b4c29c934d414cc198c9e7a8/crates/anvil/src/lib.rs#L212-L212) that we also want for reth. This `NodeHandle` is the main entrypoint to calling RPC methods in tests, or accessing spawned node components generally.

The reth node builder will start with a config, like anvil. The config can implement a `launch` method:
```rust
impl NodeConfig {
    /// Launches the node, also adding any RPC extensions passed.
    ///
    /// # Example
    /// ```rust
    /// ...omitted
    /// ```
    pub async fn launch<E: RethCliExt>(
        mut self,
        mut ext: E::Node,
        executor: TaskExecutor,
    ) -> eyre::Result<NodeHandle> {
```

A method launching a node for testing would then be simple:
```rust
/// A simple function to launch a node with the specified [NodeConfig], spawning tasks on the
/// [TaskExecutor] constructed from [tokio_runtime].
pub async fn spawn_node(config: NodeConfig) -> eyre::Result<NodeHandle> {
    let runtime = tokio_runtime()?;
    let task_manager = TaskManager::new(runtime.handle().clone());
    let ext = DefaultRethNodeCommandConfig;
    config.launch::<()>(ext, task_manager.executor()).await
}
```

Like anvil, this should also return a `NodeHandle` to be used in tests. Reth already has `RethRpcServerHandles` which provide access to normal and auth RPCs:
```rust
use reth::{cli::components::RethRpcServerHandles, rpc::api::EthApiClient};
async fn t(handles: RethRpcServerHandles) {
    let client = handles.rpc.http_client().expect("http server not started");
    let block_number = client.block_number().await.unwrap();
}
```

## Challenges
One main challenge in getting this working will be supporting `TempDatabase` alongside `DatabaseEnv`. This could be done by:
* Supporting `DB: Database` generics in all node init methods
* Creating a struct to handle initialization based on whether the user wants a real or test database
* Implementing a 2-step launch method:
```rust
/// Install a database
pub fn install_db<DB: Database>(self, database: DB) -> eyre::Result<ConfigWithDatabase<DB>> {
```
```rust
let db = create_test_rw_db();
let installed_config = config.install_db(db); // -> ConfigWithDatabase<Arc<TempDatabase<DatabaseEnv>>>
let handle = installed_config.launch::<()>(ext, task_manager.executor()).await;
```

This could also be done by creating `EitherDatabase<A, B>`, which would implement `Database`. It would not be as general as other `Either` types though, since it would need specific associated types. This is probably fine, since the trait is already `Sealed` and we only care about `DatabaseEnv` and `TempDatabase<DB>`.
```rust
impl<A: Database, B: Database> Database for EitherDatabase<A, B> {
    type TX = tx::Tx<RO>;
    type TXMut = tx::Tx<RW>;
    // ...
}
```
It's important to note that `Box<dyn Database>` cannot be used because `Database` is not object-safe.

## Usage in tests
With these components, launching a node and writing an RPC test should look something like this:
```rust
#[tokio::test]
async fn test_node_config() {
    // spawn_test_node would install a test / tmp DB
    let test_genesis = test_genesis();
    let config = NodeConfig::default().with_http().with_genesis(test_genesis);
    let handle = spawn_test_node().await;
    
    // check block number
    let block = handle.block_number().await.unwrap();

    // give reth a payload - generating these args or gathering this test data will require some more work
    handle.new_payload(/* args */).await.unwrap();
    handle.fork_choice_updated(/* args */).await.unwrap();
    
    // get next block, check increased block num
    let second_block = handle.block_number().await.unwrap();
    assert_eq!(block + 1, second_block);
}
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Rjected**
This implements the node builder designed in #5693, and adds a simple block number test with the following UX:
```rust
// this launches a test node with http
let rpc_args = RpcServerArgs::default().with_http();
let handle = spawn_node(NodeBuilder::test().with_rpc(rpc_args)).await.unwrap();

// call a function on the node
let client = handle.rpc_server_handles().rpc.http_client().unwrap();
let block_number = client.block_number().await.unwrap();

// it should be zero, since this is an ephemeral test node
assert_eq!(block_number, U256::ZERO);
```

TODO:
- [x] Fix tests - we currently install the prometheus recorder more than once if we spin up multiple nodes in a test

For follow-up:
- Provide a handle method for shutting down the node. It might also need to run when all handles are dropped?

Fixes #5693

### CODE REVIEW

File: bin/reth/src/node/mod.rs [hunk 1]
```diff
@@ -600,392 +236,11 @@ impl<Ext: RethCliExt> NodeCommand<Ext> {
...
-        self.rpc.http_port -= self.instance - 1;
-        // ws port is scaled by a factor of instance * 2
-        self.rpc.ws_port += self.instance * 2 - 2;
-    }
 }
 
 /// Drives the [NetworkManager] future until a [Shutdown](reth_tasks::shutdown::Shutdown) signal is
 /// received. If configured, this writes known peers to `persistent_peers_file` afterwards.
-async fn run_network_until_shutdown<C>(
+pub(crate) async fn run_network_until_shutdown<C>(
```
Comment (mattsse) [inline bin/reth/src/node/mod.rs:243]: we should move this somewhere else

**Review by mattsse — COMMENTED** [review]
this is great!

I like this a lot.

I have a few ideas for followups, but these will be related to better node type abstraction.

with the builder we now have the flexibility to make incremental changes.

as immediate followup, I suggest we refactor the bin/reth crate layout:

group commands in commands folder,

move NodeBuilder and related to `node`, this is prep for extracting that code to a standalone repo

File: bin/reth/src/node/mod.rs (hunk 1 above)
Comment (Rjected) [reply bin/reth/src/node/mod.rs]: added a `run_until_graceful_shutdown` method which takes a `FnOnce` to run. Writing the peers to a file is extracted into its own method, which is now in `bin/reth/src/utils.rs`.

**Review by mattsse — APPROVED** [review]
I like this a lot.

this unblocks more generalization.
we can now make more incremental changes and improve extensions and UX

**Pull request merged**